pkg/policy/groups/controllers.go

```diff
@@ -0,0 +1,49 @@
+// Copyright 2017-2018 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package groups
+
+import (
+	"time"
+
+	cilium_v2 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2"
+)
+
+const (
+	maxConcurrentUpdates = 4
+)
+
+func init() {
+	go UpgradeCNPInformation()
+}
+
+// UpgradeCNPInformation  retrieves all the CNP that has currently a derivative
+// policy and creates the new derivatives policies with the latest information
+// from providers.  To avoid issues with rate-limiting this function will
+// execute the addDerivative function with a max number of concurrent calls,
+// defined on maxConcurrentUpdates.
+func UpgradeCNPInformation() {
+	for {
+		cnpToUpdate := groupsCNPCache.GetAllCNP()
+		sem := make(chan bool, maxConcurrentUpdates)
+		for _, cnp := range cnpToUpdate {
+			sem <- true
+			go func(cnp *cilium_v2.CiliumNetworkPolicy) {
+				defer func() { <-sem }()
```

Comment:
User a `sync.WaitGroup` is less complicated to understand what you were trying to achive with the channels.